study-builder/tenants/testboston/rules/link-admin-accounts.js

```diff
@@ -0,0 +1,119 @@
+// This rule will look for admin account and link it to the Google social account. Account link is
+// determined by using the email. This is only executed for the admin client and only if the
+// accounts are not already linked.
+//
+// This was derived from Auth0's rule template from here:
+// https://github.com/auth0/rules/blob/b381ef4a1d5df16c1716ea066f46cff87b88f7f1/src/rules/link-users-by-email-with-metadata.js
+//
+// This assumes the rule runtime is Node 12, which is configured in Auth0 Dashboard Tenant Settings.
+function(user, context, callback) {
+  const _ = require('lodash');
+  const request = require('request');
+  const adminConnection = 'study-admin-authentication';
+  const googleConnection = 'google-oauth2';
+
+  // Note that client metadata key/values are all strings.
+  context.clientMetadata = context.clientMetadata || {};
+  const isAdminClient = context.clientMetadata.isAdminClient === 'true';
+
+  if (!isAdminClient || context.connection !== googleConnection) {
+    console.log('Either not admin client or not google connection, skipping account linking');
+    return callback(null, user, context);
+  }
+
+  const adminIdentity = user.identities.find(identity => identity.connection === adminConnection);
+  if (adminIdentity) {
+    console.log('Current user already has admin identity, meaning google identity has already been linked');
+    return callback(null, user, context);
+  }
+
+  request({
+    url: auth0.baseUrl + '/users',
+    headers: {
+      Authorization: 'Bearer ' + auth0.accessToken
+    },
+    qs: {
+      search_engine: 'v3',
+      q: `identities.connection:"${adminConnection}" AND email:"${user.email}"`
+    }
+  }, function(err, response, body) {
+    if (err) {
+      console.log('Error while searching for admin user with email:', user.email);
+      console.log(err);
+      return callback(err);
+    }
+
+    if (response.statusCode !== 200) {
+      console.log('Failed to search for admin user with email:', user.email);
+      console.log(body);
+      return callback(new Error(body));
+    }
+
+    const data = JSON.parse(body);
+    if (data.length !== 1) {
+      console.log('Expected to find one admin user but found', data.length);
+      return callback(new Error("Could not find a corresponding study admin account"));
```

Comment:
This means anyone trying to login will fail, unless we find them in the `study-admin-authentication` database. And users need to be manually added to that database, so this gives us fine control of who gets in.

Also, failing here will cause the auth pipeline to halt. Since this is the first rule (see `tenant.yaml`) it means we won't run the `register-user-in-pepper.js` rule. Auth0 will still create a profile for them, but at the very least, it won't hit Pepper so no users will be created in backend.

Reply:
That's great- sounds like it'll save us a few headaches later. Given we're squashing the commits together, could you add a comment noting that behavior?

Reply:
Saw this ordering clearly specified in the tenant setup file further down- my comment can be disregarded.